pkg/healthcheck/healthcheck.go

```diff
@@ -273,18 +275,7 @@ func (hc *HealthChecker) addLinkerdDataPlaneChecks() {
 		retry:       hc.ShouldRetry,
 		fatal:       false,
 		check: func() error {
-			req := &pb.ListPodsRequest{}
-			if hc.DataPlaneNamespace != "" {
-				req.Namespace = hc.DataPlaneNamespace
-			}
-			// ListPods returns all pods, but we can use the `Added` field to verify
-			// which are found in Prometheus
-			resp, err := hc.apiClient.ListPods(context.Background(), req)
-			if err != nil {
-				return err
-			}
-
-			return validateDataPlanePodReporting(hc.dataPlanePods, resp.GetPods())
+			return validateDataPlanePodReporting(hc.dataPlanePods)
```

Comment:
This check retries if the metrics aren't present yet, but without re-calling the ListPods API, there's no possibility that the check will succeed when it's retried. Instead I think we need to call `ListPods` every time this check executes.

Also, the previous list of data plane pods was already filtered to only include those pods that have been injected to point to `hc.ControlPlaneNamespace`, but the response from `ListPods` is not. I think it would be worthwhile to add a helper method that calls `ListPods` and does the filtering. Something like:

```
func (hc *HealthChecker) getDataPlanePods() ([]*pb.Pod, error) {
	req := &pb.ListPodsRequest{}
	if hc.DataPlaneNamespace != "" {
		req.Namespace = hc.DataPlaneNamespace
	}

	resp, err := hc.apiClient.ListPods(context.Background(), req)
	if err != nil {
		return nil, err
	}

	pods := make([]*pb.Pod, 0)
	for _, pod := range resp.GetPods() {
		if pod.ControllerNamespace == hc.ControlPlaneNamespace {
			pods = append(pods, pod)
		}
	}

	return pods, nil
}
```

And then in this check, you could do:

```
		check: func() error {
			pods, err := hc.getDataPlanePods()
			if err != nil {
				return err
			}

			return validateDataPlanePodReporting(pods)
		},
```